### PR TITLE
query rewrite for LogsTable skipping index

### PR DESCRIPTION
**Pull request by seankao-az**
### Description
Query rewrite for LogsTable skipping index

#### Build skipping index
Process for building skipping index is unchanged due to compatibility of current method with LogsTable.

#### Query rewrite for skipping index
On query plan optimization time, we construct a new LogsTable with the DataFrame to fetch log file ids from skipping index. These log file ids are then used to build the scan operator.

#### Dependency
Added a compile time dependency which contains only the interface of LogsTable.

#### Test
Test is not possible without loading LogsConnectorSpark fat jar as dependency. Instead, manual integration test is done locally.

### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._ 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala [hunk 1]
```diff
@@ -57,6 +62,48 @@ class ApplyFlintSparkSkippingIndex(flint: FlintSpark) extends Rule[LogicalPlan]
...
+         */
+        if (indexFilter.isDefined) {
+          val indexScan = flint.queryIndex(skippingIndex.name())
+          val selectedFiles =
+            // Non hybrid scan
+            // TODO: refactor common logic with file-based skipping index
+            indexScan
+              .filter(new Column(indexFilter.get))
+              .select(FILE_PATH_COLUMN)
+              .collect
```
[the remark this answers is not in the document]
Comment (seankao-az) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala:89]: That's right. Need changes in `LogsTable` side. Right now `LogsTable` accepts a list of file ids. Should let it accept `DataFrame` instead.

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala [hunk 2]
```diff
@@ -57,6 +62,48 @@ class ApplyFlintSparkSkippingIndex(flint: FlintSpark) extends Rule[LogicalPlan]
...
+              .map(_.getString(0))
+
+          // Construct LogsTable with selectedFiles as its log file ids
+          // It will build scan operator using these log file ids
+          val logsTable = table.asInstanceOf[LogsTable]
+          val newTable = new LogsTable(
+            logsTable.getSchema(),
+            logsTable.getOptions(),
+            selectedFiles,
+            logsTable.getProcessedFields())
```
Comment (dai-chen) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala:99]: I'm thinking should we use reflection to remove code dependency.

Comment (seankao-az) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala:99]: mark as resolved. Code dependency is just a empty implementation of LogsTable for compile

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (seankao-az) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala]: Done. Integration tested locally together with changes from dependency package.

**Review by dai-chen — APPROVED** [review]

**Review by penghuo — APPROVED** [review]

**Pull request merged**